Pilot/proxyTools.py

```
@@ -65,3 +82,96 @@ def getVO(proxy_data):
             if match:
                 return match.groups()[0].decode()
     raise NotImplementedError("Something went very wrong")
+
+
+class BaseConnectedRequest(object):
+    """This class helps supporting multiple kinds of requests that requires connections"""
+
+    def __init__(self, url, caPath, name="unknown"):
+        self.name = name
+        self.url = url
+        self.caPath = caPath
+        # We assume we have only one context, so this variable could be shared to avoid opening n times a cert.
+        # On the contrary, to avoid race conditions, we do avoid using "self.data" and "self.headers"
+        self._context = None
+
+        self._prepareRequest()
+
+    def generateUserAgent(self, pilotUUID):
+        """To analyse the traffic, we can send a taylor-made User-Agent
+
+        Args:
+            pilotUUID (str): Unique ID of the Pilot
+
+        Returns:
+            str: The generated user agent
+        """
+        return "Dirac Pilot [%s]" % pilotUUID
+
+    def _prepareRequest(self):
+        """As previously, loads the SSL certificates of the server (to avoid "unknown issuer")"""
+        # Load the SSL context
+        self._context = ssl.create_default_context()
+        self._context.load_verify_locations(capath=self.caPath)
+
+    def executeRequest(self, raw_data, headers={"User-Agent": "Dirac Pilot [Unknown ID]"}):
```

I would move `headers` as a private attribute of `BaseConnectedRequest`, to avoid repeating the default value multiple times and also because it can be misleading:

```python
In [6]: def executeRequest(headers={"User-Agent": "Dirac Pilot [Unknown ID]"}, i=1):
   ...:     headers[f"Bearer{i}"] = "coucou"
   ...:     print(headers)
   ...: 

In [7]: executeRequest(i=1)
{'User-Agent': 'Dirac Pilot [Unknown ID]', 'Bearer1': 'coucou'}

In [8]: executeRequest(i=2)
{'User-Agent': 'Dirac Pilot [Unknown ID]', 'Bearer1': 'coucou', 'Bearer2': 'coucou'}

In [9]: executeRequest(i=3)
{'User-Agent': 'Dirac Pilot [Unknown ID]', 'Bearer1': 'coucou', 'Bearer2': 'coucou', 'Bearer3': 'coucou'}
```

(You are not impacted by this issue, it's just to highlight the potential risk.)

Also you could directly update the user agent value in the headers in `generateUserAgent()`

---

I added header as a parameter to customize it easily. I do agree that it would be nice to put it as a "private" field.

In the future, if needed to add headers (for example, `Content-Type` or `Accept`), we could add a `conn.addHeader()`

---

You can make it `protected` (may be not private because you need to access it from the inheriting class) by writing `self._headers` instead of `self.headers`.